Docstring style guide (#1794)
sopel-irc/sopel: CONTRIBUTING.md

Submitting Issues
-----------------

When submitting issues to our
[issue tracker](https://github.com/sopel-irc/sopel/issues), it's important
that you do the following:

1. Describe your issue clearly and concisely.
2. Give Sopel the `.version` command, and include the output in your issue.
3. Note the OS you're running Sopel on, and how you installed Sopel (via your
   package manager, pip, `setup.py install`, or running straight from source)

> 4. Include relevant output from the log files in `~/.sopel/logs.`

Reviewer: IIRC, logs is a folder, and as such, I feel like it's better to add a trailing /

```suggestion
4. Include relevant output from the log files in `~/.sopel/logs/`.
```

I assume the trailing dot wasn't supposed to be included.

Author: Indeed, I think that was a regex replace.

  commit. Preferably, one change in functionality should be in exactly one
  commit—and commits should be atomic (i.e. try not to commit broken or
  intermediate states of the code).
* pep8ify your code before you commit. We don't worry about line length much
  (though it's good if you do keep lines short), but you should try to follow
  the rest of the rules.
* Test your code before you commit. We don't have a formal testing plan in
  place, but you should make sure your code works as promised before you
  commit.
* If you have cloned the repository to your personal computer, you can use our
  provided git hooks to automatically check whether your code style is up to
  Sopel's requirements and that tests still pass, before committing your code.
  The git hooks can be enabled by running `make install-hooks`, and disabled by
  running `make uninstall-hooks`,  in your root `sopel` folder.
* Make your commit messages clear and explicative. Our convention is to place
  the name of the thing you're changing in at the beginning of the message,
  followed by a colon: the module name for modules, docs for documentation

>   files, coretasks for coretasks.py, db for the database feature, and so on.

Reviewer: ```suggestion
  files, coretasks for `coretasks.py`, db for the database feature, and so on.
```

Author: I'm going to backtick a lot more than just the one filename. 😂

>   the name of the thing you're changing in at the beginning of the message,
>   followed by a colon: the module name for modules, docs for documentation
>   files, coretasks for coretasks.py, db for the database feature, and so on.

Author: ```suggestion
  the name of the thing you're changing in at the beginning of the message,
  followed by a colon: the plugin name for plugins, "docs" for documentation
  files, "coretasks" for `coretasks.py`, "db" for the database feature, etc.
```

Tweaks inspired by @Exirel's suggestion on line 47, but I had to integrate them to avoid a conflict when applying.

  second, if the first is `#!/usr/bin/env python`), and `from __future__ import
  unicode_literals, absolute_import, print_function, division` as the first
  line after the module docstring.

Documenting Code
----------------

Hopefully you're documenting new code as you write it. We like to use these
guidelines for writing docstrings. Let's start with an example:

```
def scramble_eggs(eggs, bacon=None, spam=False):
"""Scramble eggs, with optional bacon and/or SPAM.

:param int eggs: how many eggs to scramble
:param bacon: optional bacon to put in the ``eggs``
:type bacon: :class:`sopel.tools.breakfast.Bacon`
:param bool spam: whether to put SPAM in the scrambled ``eggs``
:return: the scrambled eggs
:rtype: :term:`iterable` of :class:`sopel.tools.breakfast.Egg`

You should create and customize your own :class:`~sopel.tools.bacon.Bacon`
object to pass in. See that class's documentation to learn how to make the
bacon extra crispy, chopped, diced, etc.

This function will generate SPAM as needed, since it is not customizable.

.. versionadded:: 7.1
.. seealso::

   To make an omelet, use :func:`make_omelet`.

"""

> # <function body>

Reviewer: The body needs a 4 spaces indentation.

Reviewer: Note: the docstring **is part** of the function's body.

Author: ```suggestion
    """Scramble eggs, with optional bacon and/or SPAM.

    :param int eggs: how many eggs to scramble
    :param bacon: optional bacon to put in the ``eggs``
    :type bacon: :class:`sopel.tools.breakfast.Bacon`
    :param bool spam: whether to put SPAM in the scrambled ``eggs``
    :return: the scrambled eggs
    :rtype: :term:`iterable` of :class:`sopel.tools.breakfast.Egg`

    You should create and customize your own :class:`~sopel.tools.bacon.Bacon`
    object to pass in. See that class's documentation to learn how to make the
    bacon extra crispy, chopped, diced, etc.

    This function will generate SPAM as needed, since it is not customizable.

    .. versionadded:: 7.1
    .. seealso::

       To make an omelet, use :func:`make_omelet`.

    """
    # <function code>
```

Fake suggestion to be applied with the rest of the batch.
